url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/protocols/specifications/TreeBasedDistanceFraudSimulator.java

```diff
@@ -0,0 +1,163 @@
+package protocols.specifications;
+
+import java.security.SecureRandom;
+import java.util.TreeMap;
+
+
+/*Trujillo- Mar 11, 2014
+ * This simulator is similar to an exhaustive search. However, it limits
+ * the number of sequences kept in memory. Instead of having all of them, 
+ * we just keep the ones that are repeated more often.*/
+public class TreeBasedDistanceFraudSimulator{
+	
+	public static final int RUNS = 100000;
+	public static final int DEFAULT_SIZE = 65536;
+
+	private int size = DEFAULT_SIZE;
+	
+	private SecureRandom random;
+			
+	public TreeBasedDistanceFraudSimulator(){			
+		random = new SecureRandom();
+	}
+	
+	public double computeDistanceFraud(int n){
+		double result = 0;
+		for (int i = 0; i < RUNS; i++) {
+			SequenceInTree mostFrequentSequence = runForMostFrequentSequence(n);
+			//System.out.println(mostFrequentSequence.toString());
+			result += (double)mostFrequentSequence.frequency/Math.pow(2, n);
+		}
+		return result/RUNS;
+	}
+	
+	private SequenceInTree runForMostFrequentSequence(int n){
+		TreeMap<SequenceInTree, SequenceInTree> sequences = new TreeMap<>();
+		SequenceInTree first = new SequenceInTree("", 1);
+		sequences.put(first, first);
+		sequences = runForMostFrequentSequence(n, sequences);
+		return sequences.firstKey();
+	}
+	
+	/*Trujillo- Mar 11, 2014
+	 * In this method, we have list of sequences and their number of appearances.
+	 * Let "s" be a sequence and "a" its number of appearances. If n > 0, both
+	 * "s0" and "s1" can be generated appearing "a0" and "a1" times respectively 
+	 * where a0 + a1 = 2*a. Therefore, a0 is "randomly" generated and a1 is computed
+	 * as the Equation before. If, for instance, a0 is already in the list, 
+	 * we sum its appearances, otherwise we add it.
+	 * 
+	 * Once all the sequences have been added, the size of the list is decreased down
+	 * to SIZE by keeping only those sequences with the higher number of
+	 * appearances.
+	 * */
+		
+	/*Trujillo- Mar 31, 2014
+	 * This in old a not efficient implementation, yet it works.*/
+	private TreeMap<SequenceInTree, SequenceInTree> runForMostFrequentSequence(int n, 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.